feat(es/minifier): Implement minifier

kdy1 opened the pull request:
# Postponed

To simplify management, I'll merge this PR as incomplete state and implement more rules with follow-up PRs.

swc_bundler:
 - [ ] Add option for minification.

swc_ecma_minifier:
 - [ ] Implement.
 - [ ] Provide js api.

soilSpoon commented:
Are you thinking of implementing it using UglifyJS?

kdy1 commented:
@soilSpoon Yes, but mainly because I don't know well about internals of minifiers.
Is there something better than uglifyjs?

Or I can add some passes after implementing it using uglifyjs.

soilSpoon commented:
[uglifyjs](https://github.com/mishoo/UglifyJS) doesn't seem to be able to minify ECMAScript 2015 or above. So, to use this, you need to use it after transpile it to ECMAScript 2015 or lower. I think this can adversely affect performance. So I look better to use [terser](https://terser.org/) that can minify ECMAScript 2015 or above.

kdy1 commented:
@soilSpoon Oh. I got it. Thanks!
I'll port terser instead.

divmain commented:
You may also want to look at [babel-minify](https://github.com/babel/minify) for inspiration.  From what I recall, it doesn't achieve quite the same size wins as Uglify or Terser.  But the port may be significantly more straightforward - the minifier is split into [discrete operations](https://github.com/babel/minify/tree/master/packages) for the different types of transformations that it applies.

kdy1 committed:
More goldens

kdy1 committed:
Update test refs

kdy1 committed:
analyzer: Mark do-while and while as loop.

kdy1 committed:
Sort methods

kdy1 committed:
More goldens

kdy1 committed:
WIP: drop_if_break

kdy1 committed:
Update test refs

kdy1 committed:
Add `as_expr` to ExprOptExt

kdy1 committed:
Add ExprExt.cast_to_number

kdy1 committed:
as_bool: Support subtraction.

kdy1 committed:
Add a ways to prepend / append statements.

kdy1 committed:
loops: Remove for statement completely

kdy1 committed:
Update test refs

kdy1 committed:
More goldens

kdy1 committed:
Doc

kdy1 committed:
WIP: while_to_for

kdy1 committed:
Convert while loops to for loops

kdy1 committed:
More goldens

kdy1 committed:
Rename

kdy1 committed:
Convert do-while loops to for loops

kdy1 committed:
Logging

kdy1 committed:
analyzer: Mark for stmt as a loop.

kdy1 committed:
More goldens

kdy1 committed:
inline: Handle very simple functions.

kdy1 committed:
More goldens

kdy1 committed:
inline: Revent simple-function inliner

kdy1 committed:
Update goldens

kdy1 committed:
loops: Remove test of a for stmt if possible.

kdy1 committed:
Update test refs

kdy1 committed:
More goldens

kdy1 committed:
impl optimize_for_init

kdy1 committed:
Fix optimize_for_init

kdy1 committed:
ignore_return_value: Drop pure calls

kdy1 committed:
Rename

kdy1 committed:
More goldens

kdy1 committed:
Update test refs

kdy1 committed:
Rename

kdy1 committed:
WIP: optimize_loops_with_break

kdy1 committed:
loops: Remove loops with instant break

kdy1 committed:
More goldens

kdy1 committed:
loops: Preserve cases if required

kdy1 committed:
More goldens

kdy1 committed:
Update test refs

kdy1 committed:
ignore_return_value: Support class expressions

kdy1 committed:
More goldens

kdy1 committed:
Don't invoke iife in obj of member expr

kdy1 committed:
fixup

kdy1 committed:
Don't inline generators.

kdy1 committed:
More goldens

kdy1 committed:
WIP: optimize_arrow_method_prop

kdy1 committed:
impl MapWithMut

kdy1 committed:
fixup

kdy1 committed:
impl optimize_arrow_method_prop

kdy1 committed:
More goldens

kdy1 committed:
Preserve directives

kdy1 committed:
More goldens

kdy1 committed:
Preserve use strict

kdy1 committed:
Preserve directives, really

kdy1 committed:
Remove use strict if it's an only item.

kdy1 committed:
Drop calls to empty arrow expressions.

kdy1 committed:
Remove `void 0;`

kdy1 committed:
More goldens

kdy1 committed:
Change pass ordering

kdy1 committed:
Check for `this` before optimizing properties.

kdy1 committed:
fixup

kdy1 committed:
Update test refs

kdy1 committed:
analyzer: Handle setter

kdy1 committed:
More goldens

kdy1 committed:
expr simplifier: Ignore lhs of assignments.

kdy1 committed:
Update test refs

kdy1 committed:
WIP

kdy1 committed:
More goldens

kdy1 committed:
Update test refs

kdy1 committed:
Small cleanup to add more codes.

kdy1 committed:
WIP: More rules for conditionals

kdy1 committed:
Merge cons and alt if only one arg differs.

kdy1 committed:
More goldens

kdy1 committed:
WIP: move_vars_in_subscopes

kdy1 committed:
impl MapWithMut

kdy1 committed:
WIP: extract_vars_in_subscopes

kdy1 committed:
fixup

kdy1 committed:
Fix extract_vars

kdy1 committed:
More goldens

kdy1 committed:
Respect `hoist_vars`

kdy1 committed:
More goldens

kdy1 committed:
Optimize negate

kdy1 committed:
fixup

kdy1 committed:
extract_vars: Preserve side effects

kdy1 committed:
Update test refs

kdy1 committed:
More goldens

kdy1 committed:
expr simplifier: Track modifications

kdy1 committed:
Fix extract_vars

kdy1 committed:
Update test refs

kdy1 committed:
Update test refs

kdy1 committed:
WIP: optimize_arrow_body

kdy1 committed:
Optimize arrow bodies

kdy1 committed:
Update test refs

kdy1 committed:
More goldens

kdy1 committed:
Add TODOs

kdy1 committed:
Fix a test

kdy1 committed:
impl MapWithMut

kdy1 committed:
Implement compress_tpl

kdy1 committed:
template literal => string literal

kdy1 committed:
More goldens

kdy1 committed:
analyzer: Handle params of arrow expr correctly.

kdy1 committed:
More goldens

kdy1 committed:
Print expected stdout

kdy1 committed:
tester: Support stdout tests

kdy1 committed:
More goldens

kdy1 committed:
Move codes

kdy1 committed:
Convert tpl to str only if it's all printable.

kdy1 committed:
Inline very simple functions.

kdy1 committed:
Inline more simple functions.

kdy1 committed:
WIP: We need comments

kdy1 committed:
Add comments to optimizer

kdy1 committed:
Support `/*#__NOINLINE__*/`.

kdy1 committed:
Add annotation

kdy1 committed:
Respect noinline annotations.

kdy1 committed:
More goldens

kdy1 committed:
Add annotation

kdy1 committed:
Don't print all comments

kdy1 committed:
Add annotation

kdy1 committed:
Handle `/** @const */`

kdy1 committed:
Update test refs

kdy1 committed:
Drop unused vars more aggresively.

kdy1 committed:
ignore_return_value: Support object.

kdy1 committed:
Update test refs.

kdy1 committed:
More goldens

kdy1 committed:
WIP: optimize_computed_prop_name_as_normal

kdy1 committed:
Optimize some computed prop names.

kdy1 committed:
More goldens

kdy1 committed:
Optimize prop names more agressively.

kdy1 committed:
Remove more blocks.

kdy1 committed:
More goldens

kdy1 committed:
Respect `arrows`

kdy1 committed:
Don't inline async functions.

kdy1 committed:
fixup

kdy1 committed:
fixup

kdy1 committed:
ignore_return_value: Support arrays.

kdy1 committed:
Log

kdy1 committed:
Store more `typeof`s.

kdy1 committed:
Update test refs

kdy1 committed:
Log

kdy1 committed:
Don't panic on option `module: bool`.

kdy1 committed:
lint

kdy1 committed:
Option: `global_defs`

kdy1 committed:
Move

kdy1 committed:
Move terser option into crate

kdy1 committed:
More options

kdy1 committed:
More work for options

kdy1 committed:
More work for options.

kdy1 committed:
fixup

kdy1 committed:
More work for options

kdy1 committed:
More works for option

kdy1 committed:
Even more options.

kdy1 committed:
Rename

kdy1 committed:
More goldens

pickfire commented on ecmascript/minifier/src/compress/optimize/dead_code.rs:
````diff
@@ -0,0 +1,75 @@
+    ///     throw x();
+    /// }
+    /// ```
+    pub(super) fn optimize_in_fn_termiation(&mut self, e: &mut Expr) {
````
Typo
```suggestion
    pub(super) fn optimize_in_fn_termination(&mut self, e: &mut Expr) {
```

kdy1 replied:
Oh, thanks!

kdy1 committed:
Rebase

kdy1 committed:
fixup

kdy1 committed:
fixup

kdy1 committed:
fixup

kdy1 committed:
Update test refs

kdy1 committed:
WIP: drop_unused_stmt_at_end_of_fn

kdy1 committed:
Remove useless `return void`s.

kdy1 committed:
fixup

kdy1 committed:
More golden

kdy1 committed:
Track more modifications

kdy1 committed:
ignored.txt

kdy1 committed:
Add a script to ignore tests

kdy1 committed:
Ignore some

kdy1 committed:
Respect ignored.txt

kdy1 committed:
Ignore more

kdy1 committed:
Update test refs

kdy1 committed:
Update test refs

kdy1 committed:
Udpate test refs

kdy1 committed:
More ignore

kdy1 committed:
Fix the test runner

kdy1 committed:
More ignore

kdy1 committed:
Fix ignore_return_value for object literals.

kdy1 committed:
More ignore

kdy1 committed:
Update test refs

kdy1 committed:
More ignore

kdy1 committed:
unused: Preserve decls named `argument`.

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
Update test refs

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
Udpate test refs

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
Update test refs

kdy1 committed:
Update test refs

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
Fix version issue

kdy1 committed:
Rebase

kdy1 committed:
Don't inline regex

kdy1 committed:
fixup

kdy1 committed:
Update the list

kdy1 committed:
Ignore

kdy1 committed:
Don't inline `arguments`.

kdy1 committed:
Golden

kdy1 committed:
fixup: Use typeof

kdy1 committed:
fix

kdy1 committed:
Revert "fix"

This reverts commit d55b758ca410f287a569a6643f83b61dcf486d29.

kdy1 committed:
Ignore

kdy1 committed:
More ignore

kdy1 committed:
More ignore

kdy1 committed:
More ignore

fabiosantoscode commented:
Hey, I maintain Terser, and it's helplessly slow. Recently I started a port to rust (in a very early stage) but I don't think it's useful to keep two separate projects that are essentially the same. How can I help?

kdy1 commented:
@fabiosantoscode I implemented many rules but there are many rules left.
If you don't mind implementing it as a part of swc, I will add lots of comment how the minifier works and about some design decisions.

Note that being in-tree would make version management much simpler, as bundler depends on minifier and minifier depends on ast / visitor while those crates live in a momorepo.

fabiosantoscode commented:
Ok, I'll go on your fork and send PRs when I have the time.

I'm interested in how to implement `reduce_vars`, `unused`, and `inline`, because in Terser they're kind of the same for speed reasons. They're also very relevant for tree shaking.

kdy1 committed:
Add some comments

fabiosantoscode committed:
WIP: property mangler

I've started to implement the property mangler and thought it'd be nice to open a PR.

I'll probably need to start using JsWord instead of regular strings, but sometimes we're mangling strings as well.

I started with keep_quoted: 'strict' which is the most sane option, and also the simplest. I wonder if we need to have all the others as well.

fabiosantoscode committed:
use JsWord

fabiosantoscode committed:
reorder structs to match operation order

fabiosantoscode committed:
do not mangle undeclared

fabiosantoscode committed:
fix tests

fabiosantoscode committed:
mangle shorthand object props

fabiosantoscode committed:
create script to update JS properties; move domprops in

fabiosantoscode committed:
use domprops and jsprops

fabiosantoscode committed:
fix mangling defineProperty string args

Austaras commented:
Any possible to expose as webpack plugin like [terser](https://github.com/webpack-contrib/terser-webpack-plugin)

kdy1 commented:
@Austaras Of course. I will do so.

fabiosantoscode committed:
fix sorting differences between Linux and Mac

fabiosantoscode committed:
mangle private properties, refactor

fabiosantoscode committed:
Add regex option to limit mangling to props that match the regex

fabiosantoscode committed:
add a golden test

fabiosantoscode committed:
fix fmt error

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/swc-project/swc?pullRequest=1302) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/swc-project/swc?pullRequest=1302) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: kdy1<br/>:x: fabiosantoscode<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/swc-project/swc?pullRequest=1302) it.</sub>

kdy1 committed:
fixup for the rebase

fabiosantoscode committed:
make test scripts faster when the runner wants to filter tests

When developing the property mangler, I'd commented out the lines in this 'if' statement so I could run tests faster.

So I came up with this PR to enable us to filter tests when desired, and run them quickly, but also allow to run everything by omitting filter arguments.

If you have a better idea for testing quickly please let me know :) this is still pretty slow.

kdy1 committed:
fixup for rebase

kdy1 committed:
fixup

fabiosantoscode committed:
implement timings struct

kdy1 commented:
@fabiosantoscode

How do you think about merging this into master and improving it gradually, with multiple PRs?

fabiosantoscode commented:
Sure! Especially if it's documented as "beta"

fabiosantoscode commented:
Eventually I'd like to test this against every single Terser test. Not all of them require exact output. Some of them run the resulting code using node's VM module and compare the result.

kdy1 committed:
Make it beta

kdy1 committed:
Fix

kdy1 committed:
Fix

kdy1 committed:
More fix

kdy1 committed:
Use master

kdy1 committed:
fixup

kdy1 committed:
fixup

kdy1 committed:
Use master

kdy1 committed:
fixup

kdy1 committed:
Update tets refs

kdy1 committed:
Use master

kdy1 committed:
lint

kdy1 committed:
Ignore

kdy1 commented:
> Eventually I'd like to test this against every single Terser test. Not all of them require exact output. Some of them run the resulting code using node's VM module and compare the result.

👍  That's also my goal.
Terser has strength in the output and I want to make it faster without degrading output.

SoloJiang commented:
However, I see this PR had beed merged, but how can I try it out? At now, it only removes comment and space in my code.

kdy1 commented:
@SoloJiang It's not published yet because it's not done

SoloJiang commented:
Okay, I'm working on use swc instead of Babel and Terser in our project github.com/alibaba/ice. In our situation, bundle size is more important than bundler speed. Without minify, swc result is bigger 10% than Babel. In minify, it's 200% than Terser or esbuild which optimized by comfused code.

fabiosantoscode commented:
So much excitement for this port, I love it 💖

I need to try to reduce my hours in my side gig so I can give this more attention.

There's some transforms Terser itself doesn't do that are feasible here (such as a dictionary for common immutable constants or cross-module intelligence used in combination with spack) that can reduce this port's output bytes to a point that makes it a viable alternative to Terser.

kdy1 commented:
I'm glad to hear that and thank you!


> There's some transforms Terser itself doesn't do that are feasible here (such as a dictionary for common immutable constants or cross-module intelligence used in combination with spack) that can reduce this port's output bytes to a point that makes it a viable alternative to Terser.

Sounds interesting. I expect such cross-module transforms to reduce final output by a margin because libraries used a lot along with UI frameworks like react encourages such code.
